commons-numbers-examples/examples-jmh/src/main/java/org/apache/commons/numbers/examples/jmh/core/EuclideanNormAlgorithms.java

```diff
@@ -0,0 +1,751 @@
+        private static final double SCALE_DOWN = 0x1.0p-600;
+
+        /** Value used to scale up small numbers. */
+        private static final double SCALE_UP = 0x1.0p+600;
+
+        /** {@inheritDoc} */
+        @Override
+        public double applyAsDouble(final double[] v) {
+         // Sum of big, normal and small numbers with 2-fold extended precision summation
+            double s1 = 0;
+            double s2 = 0;
+            double s3 = 0;
+            double c1 = 0;
+            double c2 = 0;
+            double c3 = 0;
+            for (int i = 0; i < v.length; i++) {
+                final double x = Math.abs(v[i]);
+                if (!(x <= Double.MAX_VALUE)) {
+                    return x;
+                } else if (x > LARGE_THRESH) {
+                    // Scale down big numbers
+                    final double y = square(x * SCALE_DOWN);
+                    final double t = s1 + y;
+                    c1 += DoublePrecision.twoSumLow(s1, y, t);
+                    s1 = t;
+                } else if (x < SMALL_THRESH) {
+                    // Scale up small numbers
+                    final double y = square(x * SCALE_UP);
+                    final double t = s3 + y;
+                    c3 += DoublePrecision.twoSumLow(s3, y, t);
+                    s3 = t;
+                } else {
+                    // Unscaled
+                    final double y = square(x);
+                    final double t = s2 + y;
+                    c2 += DoublePrecision.twoSumLow(s2, y, t);
+                    s2 = t;
+                }
+            }
+            // The highest sum is the significant component. Add the next significant.
+            // Adapted from LinearCombination dot2s summation.
+            if (s1 != 0) {
+                s2 = s2 * SCALE_DOWN * SCALE_DOWN;
+                c2 = c2 * SCALE_DOWN * SCALE_DOWN;
+                double sum = s1 + s2;
+                c1 += DoublePrecision.twoSumLow(s1, s2, sum) + c2;
+                return Math.sqrt(sum + c1) * SCALE_UP;
+            } else if (s2 != 0) {
+                s3 = s3 * SCALE_DOWN * SCALE_DOWN;
+                c3 = c3 * SCALE_DOWN * SCALE_DOWN;
+                double sum = s2 + s3;
+                c2 += DoublePrecision.twoSumLow(s2, s3, sum) + c3;
+                return Math.sqrt(sum + c2);
+            }
+            return Math.sqrt(s3) * SCALE_DOWN;
+        }
+
+        /** Compute the square of {@code x}.
+         * @param x input value
+         * @return square of {@code x}
+         */
+        private static double square(final double x) {
+            return x * x;
+        }
+    }
+
+    /** Euclidean norm computation algorithm that uses {@link LinearCombinations} to perform
+     * an extended precision summation.
+     */
+    static final class ExtendedPrecisionLinearCombination implements ToDoubleFunction<double[]> {
+
+        /** {@inheritDoc} */
+        @Override
+        public double applyAsDouble(final double[] v) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 360-375 into a separate method.